Commit ID: 93dff4392b0a01efc2ebfc71feb40838e856a0b3
Change ID: xvxrvspymotrxuymnqrrlmposvnvqppu
Author   : Natanael Copa <ncopa@alpinelinux.org> (2022-11-16 11:55:34)
Committer: flukey <flukey@vapourmail.eu> (2026-10-05 15:58:35)

    lvresize: use POSIX shell

diff --git a/scripts/lvresize_fs_helper.sh b/scripts/lvresize_fs_helper.sh
index 7d64409a62..3b7982f610 100755
--- a/scripts/lvresize_fs_helper.sh
+++ b/scripts/lvresize_fs_helper.sh
@@ -1,4 +1,4 @@
-#!/bin/bash
+#!/bin/sh
 #
 # Copyright (C) 2022-2025 Red Hat, Inc. All rights reserved.
 #
@@ -12,10 +12,10 @@
 # along with this program; if not, write to the Free Software Foundation,
 # Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
 
-set -euE -o pipefail
+set -eu
 
-PATH="/sbin:/usr/sbin:/bin:/usr/bin:$PATH"
-GETOPT="getopt"
+PATH="/usr/bin:$PATH"
+GETOPT="ugetopt"
 SCRIPTNAME=$(basename "$0")
 DM_DEV_DIR="${DM_DEV_DIR:-/dev}"
 
@@ -96,7 +96,7 @@
 # 4 = errors left uncorrected, 8 = operational error
 # 16 = usage error, 32 = canceled, 128 = shared library error
 accept_e2fsck() {
-	local ret=0
+	ret=0
 	"$@" || ret=$?
 
 	case "$ret" in
@@ -140,17 +140,22 @@
 }
 
 btrfs_path_major_minor() {
-	local STAT
+	STAT=""
 
-	STAT=$(stat --format "echo \$((0x%t)):\$((0x%T))" "$(readlink -e "$1")") || \
+	STAT=$(stat -f "echo \$((0x%XHr)):\$((0x%XLr))" "$(readlink -e "$1")") || \
 		errorexit "Cannot get major:minor for \"$1\"."
 	eval "$STAT"
 }
 
 btrfs_devid() {
-	local devpath=$1
-	local devid devinfo major_minor path_major_minor
-	local IFS=$'\n'
+	devpath="$1"
+	devid=
+	devinfo=
+	major_minor=
+	path_major_minor=
+	oldifs=$IFS
+	IFS='
+'
 
 	major_minor=$(btrfs_path_major_minor "$devpath")
 
@@ -168,12 +173,14 @@
 			# trim all prefix and postfix spaces from devid
 			devid=${devid#"${devid%%[![:space:]]*}"}
 			echo "${devid%"${devid##*[![:space:]]}"}"
+			IFS=$oldifs
 			return 0
 			;;
 		esac
 	done
 
 	# fail, devid not found
+	IFS=$oldifs
 	return 1
 }
 
@@ -184,10 +191,12 @@
 
 # Function to detect XFS mount options
 detect_xfs_mount_options() {
-	local device=$1
-	local qflags_output qflags_hex
-	local prefix acct_flag enfd_flag
-	local -a opts=()
+	device=$1
+	qflags_output=
+	qflags_hex=
+	prefix=
+	acct_flag=
+	enfd_flag=
 	MOUNT_OPTIONS=""
 
 	# Get quota flags using xfs_db.
@@ -200,35 +209,34 @@
 	qflags_hex="${qflags_output#qflags = }"
 
 	# No flags set, no extra mount options needed.
-	[[ "$qflags_hex" == "0" ]] && return 0
+	[ "$qflags_hex" = "0" ] && return 0
 
-	if [[ ! "$qflags_hex" =~ ^0x[0-9a-fA-F]+$ ]]; then
-		logerror "xfs_db unexpected output for \"$device\": got \"$qflags_hex\""
-		return 1
-	fi
+	case $qflags_hex in
+		0[xX][0-9a-fA-F]*)
+			;;
+		*)
+			logerror "xfs_db unexpected output for \"$device\": got \"$qflags_hex\""
+			return 1
+			;;
+	esac
 
-	# Check XFS quota flags and build mount options
-	# The quota flags as defined in Linux kernel source: fs/xfs/libxfs/xfs_log_format.h:
-	#   XFS_UQUOTA_ACCT = 0x0001, XFS_UQUOTA_ENFD = 0x0002
-	#   XFS_GQUOTA_ACCT = 0x0040, XFS_GQUOTA_ENFD = 0x0080
-	#   XFS_PQUOTA_ACCT = 0x0008, XFS_PQUOTA_ENFD = 0x0200
-	#
-	# Format: "prefix acct_flag enfd_flag"
-	for quota_type in "u 0x0001 0x0002" "g 0x0040 0x0080" "p 0x0008 0x0200"; do
-		read -r prefix acct_flag enfd_flag <<< "$quota_type"
+	# Build mount options. Format: "prefix acct_flag enfd_flag".
+    # Flags as defined in fs/xfs/libxfs/xfs_log_format.h
+	while read -r prefix acct_flag enfd_flag; do
 		if [ $((qflags_hex & acct_flag)) -ne 0 ]; then
 			if [ $((qflags_hex & enfd_flag)) -ne 0 ]; then
-				opts+=("${prefix}quota")
+				MOUNT_OPTIONS="${MOUNT_OPTIONS:+${MOUNT_OPTIONS},}${prefix}quota"
 			else
-				opts+=("${prefix}qnoenforce")
+				MOUNT_OPTIONS="${MOUNT_OPTIONS:+${MOUNT_OPTIONS},}${prefix}qnoenforce"
 			fi
 		fi
-	done
-
-	# Join array elements with commas
-	MOUNT_OPTIONS=$(IFS=,; echo "${opts[*]}")
-
-	[[ -n "$MOUNT_OPTIONS" ]] && logmsg "mount options for xfs: ${MOUNT_OPTIONS}"
+	done <<'EOF'
+u 0x0001 0x0002
+g 0x0040 0x0080
+p 0x0008 0x0200
+EOF
+
+	[ -n "$MOUNT_OPTIONS" ] && logmsg "mount options for xfs: ${MOUNT_OPTIONS}"
 }
 
 fsextend() {
@@ -243,10 +251,10 @@
 	fi
 
 	if [ "$DO_FSCK" -eq 1 ]; then
-		if [[ "$FSTYPE" == "ext"* ]]; then
+	    if [ "${FSTYPE#ext}" != "$FSTYPE" ]; then
 			logmsg "e2fsck ${DEVPATH}"
 			accept_e2fsck e2fsck -f -p "$DEVPATH"
-		elif [[ "$FSTYPE" == "btrfs" ]]; then
+		elif [ "$FSTYPE" = "btrfs" ]; then
 			logmsg "btrfs check ${DEVPATH}"
 			if btrfs check "$DEVPATH"; then
 				logmsg "btrfs check done"
@@ -268,7 +276,7 @@
 	fi
 
 	if [ "$DO_MOUNT" -eq 1 ]; then
-		if [[ "$FSTYPE" == "xfs" ]]; then
+		if [ "$FSTYPE" = "xfs" ]; then
 			detect_xfs_mount_options "$DEVPATH" || logmsg "not using XFS mount options"
 		fi
 
@@ -282,7 +290,7 @@
 		fi
 	fi
 
-	if [[ "$FSTYPE" == "ext"* ]]; then
+	if [ "${FSTYPE#ext}" != "$FSTYPE" ]; then
 		logmsg "resize2fs ${DEVPATH}"
 		if resize2fs "$DEVPATH"; then
 			logmsg "resize2fs done"
@@ -290,7 +298,7 @@
 			logerror "resize2fs failed on \"$DEVPATH\""
 			RESIZEFS_FAILED=1
 		fi
-	elif [[ "$FSTYPE" == "xfs" ]]; then
+	elif [ "$FSTYPE" = "xfs" ]; then
 		logmsg "xfs_growfs ${DEVPATH}"
 		if xfs_growfs "$DEVPATH"; then
 			logmsg "xfs_growfs done"
@@ -298,7 +306,7 @@
 			logerror "xfs_growfs failed on \"$DEVPATH\""
 			RESIZEFS_FAILED=1
 		fi
-	elif [[ "$FSTYPE" == "btrfs" ]]; then
+	elif [ "$FSTYPE" = "btrfs" ]; then
 		NEWSIZEBYTES=${NEWSIZEBYTES:-max}
 		BTRFS_DEVID="$(btrfs_devid "$DEVPATH")"
 		REAL_MOUNTPOINT="$MOUNTDIR"
@@ -331,8 +339,8 @@
 
 	# If the fs was temporarily unmounted, now remount it.
 	# Not considered a command failure if this fails.
-	if [[ $DO_UNMOUNT -eq 1 && $REMOUNT -eq 1 ]]; then
-		if [[ "$FSTYPE" == "xfs" ]]; then
+	if [ $DO_UNMOUNT -eq 1 ] && [ $REMOUNT -eq 1 ]; then
+		if [ "$FSTYPE" = "xfs" ]; then
 			detect_xfs_mount_options "$DEVPATH" || logmsg "not using XFS mount options"
 		fi
 
@@ -364,10 +372,10 @@
 	fi
 
 	if [ "$DO_FSCK" -eq 1 ]; then
-		if [[ "$FSTYPE" == "ext"* ]]; then
+	    if [ "${FSTYPE#ext}" != "$FSTYPE" ]; then
 			logmsg "e2fsck ${DEVPATH}"
 			accept_e2fsck e2fsck -f -p "$DEVPATH"
-		elif [[ "$FSTYPE" == "btrfs" ]]; then
+		elif [ "$FSTYPE" = "btrfs" ]; then
 			logmsg "btrfs check ${DEVPATH}"
 			if btrfs check "$DEVPATH"; then
 				logmsg "btrfs check done"
@@ -389,7 +397,7 @@
 		fi
 	fi
 
-	if [[ "$FSTYPE" == "ext"* ]]; then
+	if [ "${FSTYPE#ext}" != "$FSTYPE" ]; then
 		NEWSIZEKB=$(( NEWSIZEBYTES / 1024 ))
 		logmsg "resize2fs ${DEVPATH} ${NEWSIZEKB}k"
 		if resize2fs "$DEVPATH" "$NEWSIZEKB"k; then
@@ -399,7 +407,7 @@
 			# will exit after cleanup unmount
 			RESIZEFS_FAILED=1
 		fi
-	elif [[ "$FSTYPE" == "btrfs" ]]; then
+	elif [ "$FSTYPE" = "btrfs" ]; then
 		BTRFS_DEVID="$(btrfs_devid "$DEVPATH")"
 		REAL_MOUNTPOINT="$MOUNTDIR"
 
@@ -447,7 +455,7 @@
 
 	# If the fs was temporarily unmounted, now remount it.
 	# Not considered a command failure if this fails.
-	if [[ $DO_UNMOUNT -eq 1 && $REMOUNT -eq 1 ]]; then
+	if [ "$DO_UNMOUNT" -eq 1 ] && [ "$REMOUNT" -eq 1 ]; then
 		logmsg "remount ${DEVPATH} ${MOUNTDIR}"
 		if mount -t "$FSTYPE" "$DEVPATH" "$MOUNTDIR"; then
 			logmsg "remount done"
@@ -519,7 +527,7 @@
 	shift
 done
 
-if [ "$UID" != 0 ] && [ "$EUID" != 0 ]; then
+if [ "$(id -u)" != 0 ]; then
 	errorexit "${SCRIPTNAME} must be run as root."
 fi
 
@@ -528,11 +536,11 @@
 #
 
 # There are three top level commands: --fsextend, --fsreduce, --cryptresize.
-if [[ "$DO_FSEXTEND" -eq 0 && "$DO_FSREDUCE" -eq 0 && "$DO_CRYPTRESIZE" -eq 0 ]]; then
+if [ "$DO_FSEXTEND" -eq 0 ] && [ "$DO_FSREDUCE" -eq 0 ]  && [ "$DO_CRYPTRESIZE" -eq 0 ]; then
 	errorexit "Missing --fsextend|--fsreduce|--cryptresize."
 fi
 
-if [[ "$DO_FSEXTEND" -eq 1 || "$DO_FSREDUCE" -eq 1 ]]; then
+if [ "$DO_FSEXTEND" -eq 1 ] || [ "$DO_FSREDUCE" -eq 1 ]; then
 	case "$FSTYPE" in
 	  ext[234]) ;;
 	  "xfs")    ;;
@@ -545,7 +553,7 @@
 	fi
 fi
 
-if [[ "$DO_CRYPTRESIZE" -eq 1 && -z "$CRYPTPATH" ]]; then
+if [ "$DO_CRYPTRESIZE" -eq 1 ] && [ -z "$CRYPTPATH" ]; then
 	errorexit "Missing required --cryptpath for --cryptresize."
 fi
 
@@ -563,25 +571,25 @@
 	errorexit "Device does not exist \"$DEVPATH\"."
 fi
 
-if [[ "$DO_UNMOUNT" -eq 1 && -z "$MOUNTDIR" ]]; then
+if [ "$DO_UNMOUNT" -eq 1 ] && [ -z "$MOUNTDIR" ]; then
 	errorexit "Missing required --mountdir for --unmount."
 fi
 
-if [[ "$DO_FSREDUCE" -eq 1 && "$FSTYPE" == "xfs" ]]; then
+if [ "$DO_FSREDUCE" -eq 1 ] && [ "$FSTYPE" = "xfs" ]; then
 	errorexit "Cannot reduce xfs."
 fi
 
-if [[ "$DO_FSCK" -eq 1 && "$FSTYPE" == "xfs" ]]; then
+if [ "$DO_FSCK" -eq 1 ] && [ "$FSTYPE" = "xfs" ]; then
 	errorexit "Cannot use --fsck with xfs."
 fi
 
 if [ "$DO_MOUNT" -eq 1 ]; then
-	TMPDIR=$(mktemp --suffix _lvresize_$$ -d -p /tmp)
+    TMPDIR=$(mktemp -d -p /tmp '_lvresize_.XXXXXX')
 	if [ ! -e "$TMPDIR" ]; then
 		errorexit "Failed to create temp dir."
 	fi
 	# In case the script terminates without doing cleanup
-	function finish {
+	finish() {
 		if [ "$TMP_MOUNT_DONE" -eq 1 ]; then
 			logmsg "exit unmount ${TMPDIR}"
 			umount "$TMPDIR"
